File: backend/api/routes/memory.py
"""
记忆功能 API 路由

提供记忆的增删改查、设置管理接口。
"""

from fastapi import APIRouter, Depends
from loguru import logger

from api.deps import CurrentUserId, Database
from core.exceptions import AppException
from schemas.memory import (
    MemoryAddRequest,
    MemoryAddResponse,
    MemoryDeleteAllResponse,
    MemoryDeleteResponse,
    MemoryListResponse,
    MemorySettingsResponse,
    MemorySettingsUpdateRequest,
    MemoryUpdateRequest,
    MemoryUpdateResponse,
)
from services.memory_service import MemoryService

router = APIRouter(prefix="/memories", tags=["记忆"])


def get_memory_service(db: Database) -> MemoryService:
    """获取记忆服务实例"""
    return MemoryService(db)


# ===== 记忆设置 =====


@router.get("/settings", response_model=MemorySettingsResponse, summary="获取记忆设置")
async def get_memory_settings(
    current_user_id: CurrentUserId,
    service: MemoryService = Depends(get_memory_service),
):
    """获取当前用户的记忆功能设置"""
    try:
        return await service.get_settings(current_user_id)
    except AppException:
        raise
    except Exception as e:
        logger.error(
            f"Error in get_memory_settings | user_id={current_user_id} | error={e}"
        )
        raise AppException(
            code="ROUTE_GET_MEMORY_SETTINGS_ERROR",
            message="获取记忆设置失败",
            status_code=500,
        )


@router.put("/settings", response_model=MemorySettingsResponse, summary="更新记忆设置")
async def update_memory_settings(
    body: MemorySettingsUpdateRequest,
    current_user_id: CurrentUserId,
    service: MemoryService = Depends(get_memory_service),
):
    """更新当前用户的记忆功能设置（开关、保留天数）"""
    try:
        return await service.update_settings(
            current_user_id,
            memory_enabled=body.memory_enabled,
            retention_days=body.retention_days,
        )
    except AppException:
        raise
    except Exception as e:
        logger.error(
            f"Error in update_memory_settings | user_id={current_user_id} | error={e}"
        )
        raise AppException(
            code="ROUTE_UPDATE_MEMORY_SETTINGS_ERROR",
            message="更新记忆设置失败",
            status_code=500,
        )


# ===== 记忆 CRUD =====


@router.get("", response_model=MemoryListResponse, summary="获取记忆列表")
async def get_memories(
    current_user_id: CurrentUserId,
    service: MemoryService = Depends(get_memory_service),
):
    """获取当前用户的所有记忆"""
    try:
        memories = await service.get_all_memories(current_user_id)
        return {"memories": memories, "total": len(memories)}
    except AppException:
        raise
    except Exception as e:
        logger.error(
            f"Error in get_memories | user_id={current_user_id} | error={e}"
        )
        raise AppException(
            code="ROUTE_GET_MEMORIES_ERROR",
            message="获取记忆列表失败",
            status_code=500,
        )


@router.post("", response_model=MemoryAddResponse, summary="添加记忆")
async def add_memory(
    body: MemoryAddRequest,
    current_user_id: CurrentUserId,
    service: MemoryService = Depends(get_memory_service),
):
    """手动添加一条记忆"""
    try:
        memories = await service.add_memory(
            user_id=current_user_id,
            content=body.content,
            source="manual",
        )
        return {"memories": memories, "count": len(memories)}
    except AppException:
        raise
    except Exception as e:
        logger.error(
            f"Error in add_memory | user_id={current_user_id} | error={e}"
        )
        raise AppException(
            code="ROUTE_ADD_MEMORY_ERROR",
            message="添加记忆失败",
            status_code=500,
        )


@router.put("/{memory_id}", response_model=MemoryUpdateResponse, summary="更新记忆")
async def update_memory(
    memory_id: str,
    body: MemoryUpdateRequest,
    current_user_id: CurrentUserId,
    service: MemoryService = Depends(get_memory_service),
):
    """更新一条记忆的内容"""
    try:
        return await service.update_memory(
            memory_id=memory_id,
            content=body.content,
            user_id=current_user_id,
        )
    except AppException:
        raise
    except Exception as e:
        logger.error(
            f"Error in update_memory | user_id={current_user_id} | "
            f"memory_id={memory_id} | error={e}"
        )
        raise AppException(
            code="ROUTE_UPDATE_MEMORY_ERROR",
            message="更新记忆失败",
            status_code=500,
        )


@router.delete("/{memory_id}", response_model=MemoryDeleteResponse, summary="删除记忆")
async def delete_memory(
    memory_id: str,
    current_user_id: CurrentUserId,
    service: MemoryService = Depends(get_memory_service),
):
    """删除一条记忆"""
    try:
        await service.delete_memory(memory_id=memory_id, user_id=current_user_id)
        return {"message": "记忆已删除"}
    except AppException:
        raise
    except Exception as e:
        logger.error(
            f"Error in delete_memory | user_id={current_user_id} | "
            f"memory_id={memory_id} | error={e}"
        )
        raise AppException(
            code="ROUTE_DELETE_MEMORY_ERROR",
            message="删除记忆失败",
            status_code=500,
        )


@router.delete("", response_model=MemoryDeleteAllResponse, summary="清空所有记忆")
async def delete_all_memories(
    current_user_id: CurrentUserId,
    service: MemoryService = Depends(get_memory_service),
):
    """清空当前用户的所有记忆"""
    try:
        await service.delete_all_memories(current_user_id)
        return {"message": "所有记忆已清空"}
    except AppException:
        raise
    except Exception as e:
        logger.error(
            f"Error in delete_all_memories | user_id={current_user_id} | error={e}"
        )
        raise AppException(
            code="ROUTE_DELETE_ALL_MEMORIES_ERROR",
            message="清空记忆失败",
            status_code=500,
        )

File: frontend/src/services/memory.ts
/**
 * 记忆功能 API 服务
 */

import { request } from './api';

// ============================================================
// 类型定义
// ============================================================

/** 记忆元数据 */
export interface MemoryMetadata {
  source: 'auto' | 'manual';
  conversation_id: string | null;
}

/** 单条记忆 */
export interface MemoryItem {
  id: string;
  memory: string;
  metadata: MemoryMetadata;
  created_at: string | null;
  updated_at: string | null;
}

/** 记忆列表响应 */
export interface MemoryListResponse {
  memories: MemoryItem[];
  total: number;
}

/** 添加记忆响应 */
export interface MemoryAddResponse {
  memories: MemoryItem[];
  count: number;
}

/** 更新记忆响应 */
export interface MemoryUpdateResponse {
  id: string;
  memory: string;
  updated_at: string | null;
}

/** 记忆设置 */
export interface MemorySettings {
  memory_enabled: boolean;
  retention_days: number;
  updated_at: string | null;
}

// ============================================================
// API 函数
// ============================================================

/**
 * 获取记忆列表
 */
export async function getMemories(): Promise<MemoryListResponse> {
  return request<MemoryListResponse>({
    method: 'GET',
    url: '/memories',
    timeout: 60000, // Mem0 pgvector 查询可能较慢
  });
}

/**
 * 添加记忆
 */
export async function addMemory(content: string): Promise<MemoryAddResponse> {
  return request<MemoryAddResponse>({
    method: 'POST',
    url: '/memories',
    data: { content },
    timeout: 60000, // Mem0 LLM 提取 + 嵌入生成较慢
  });
}

/**
 * 更新记忆
 */
export async function updateMemory(
  id: string,
  content: string
): Promise<MemoryUpdateResponse> {
  return request<MemoryUpdateResponse>({
    method: 'PUT',
    url: `/memories/${id}`,
    data: { content },
    timeout: 60000,
  });
}

/**
 * 删除记忆
 */
export async function deleteMemory(id: string): Promise<void> {
  return request<void>({
    method: 'DELETE',
    url: `/memories/${id}`,
    timeout: 60000,
  });
}

/**
 * 清空所有记忆
 */
export async function deleteAllMemories(): Promise<void> {
  return request<void>({
    method: 'DELETE',
    url: '/memories',
    timeout: 60000,
  });
}

/**
 * 获取记忆设置
 */
export async function getMemorySettings(): Promise<MemorySettings> {
  return request<MemorySettings>({
    method: 'GET',
    url: '/memories/settings',
    timeout: 60000,
  });
}

/**
 * 更新记忆设置
 */
export async function updateMemorySettings(
  data: Partial<Pick<MemorySettings, 'memory_enabled' | 'retention_days'>>
): Promise<MemorySettings> {
  return request<MemorySettings>({
    method: 'PUT',
    url: '/memories/settings',
    data,
    timeout: 60000,
  });
}
